python/ray/dataframe/dataframe.py

```diff
@@ -1143,21 +1146,141 @@ def boxplot(self, column=None, by=None, ax=None, fontsize=None, rot=0,
+            return self._clip_with_scalar(
+                lower, upper, inplace=inplace
+            )
+        result = self
+        if lower is not None:
+            result = result.clip_lower(lower, axis, inplace=inplace)
+        if upper is not None:
+            if inplace:
+                result = self
+            result = result.clip_upper(upper, axis, inplace=inplace)
+
+        return result
+
     def clip(self, lower=None, upper=None, axis=None, inplace=False, *args,
              **kwargs):
-        raise NotImplementedError(
-            "To contribute to Pandas on Ray, please visit "
-            "github.com/ray-project/ray.")
+        # TODO: throw any pandas errors by validating with a dummy dataframe
+        if isinstance(self, ABCPanel):
+            raise NotImplementedError("clip is not supported yet for panels")
+
+        inplace = validate_bool_kwarg(inplace, 'inplace')
+
+        axis = nv.validate_clip_with_axis(axis, args, kwargs)
+
+        # GH 17276
+        # numpy doesn't like NaN as a clip value
+        # so ignore
+        if np.any(pd.isnull(lower)):
+            lower = None
+        if np.any(pd.isnull(upper)):
+            upper = None
+
+        # GH 2747 (arguments were reversed)
+        if lower is not None and upper is not None:
+            if is_scalar(lower) and is_scalar(upper):
+                lower, upper = min(lower, upper), max(lower, upper)
+
+        return self._clip_helper(lower=lower, upper=upper, axis=axis,
+                                 inplace=inplace)
 
     def clip_lower(self, threshold, axis=None, inplace=False):
-        raise NotImplementedError(
-            "To contribute to Pandas on Ray, please visit "
-            "github.com/ray-project/ray.")
+        if axis is not None:
+            axis = pd.DataFrame()._get_axis_number(axis)
+        return self._clip_list_like_helper(
+            threshold,
+            lambda _threshold, df: df.clip_lower(
+                _threshold, axis=axis, inplace=False),
+            axis=axis, inplace=inplace
+        )
 
     def clip_upper(self, threshold, axis=None, inplace=False):
-        raise NotImplementedError(
-            "To contribute to Pandas on Ray, please visit "
-            "github.com/ray-project/ray.")
+        if axis is not None:
+            axis = pd.DataFrame()._get_axis_number(axis)
+        # self._clip_validate(threshold, axis)
+        return self._clip_list_like_helper(
+            threshold,
+            lambda _threshold, df: df.clip_upper(
+                _threshold, axis=axis, inplace=False),
+            axis=axis, inplace=inplace
+        )
+
+    def _clip_validate(self, threshold, axis):
+        if axis is None:
+            dummy_frame = pd.DataFrame(columns=self.columns)
+            dummy_frame.clip_lower(threshold, axis)
+            dummy_frame = pd.DataFrame(index=self.index)
+            dummy_frame.clip_lower(threshold, axis)
```

Comment:
Can be done in a single `dummy_frame`

Reply:
It can't be done in a single `dummy_frame` because Pandas will fill all values with `NaN` if you give both `index` and `columns`. That can make a huge DataFrame.